torchao/quantization/quant_primitives.py

```diff
@@ -796,78 +793,81 @@ def get_group_qparams_symmetric(w, n_bit=4, groupsize=128, precision=torch.float
+    w,
+    n_bit=4,
+    group_size=128,
+    precision=torch.float32,
+):
+    scales, zeros = get_group_qparams_symmetric(w, n_bit, group_size, precision)
+    n_bit = 4
+    max_int = 2 ** (n_bit - 1) - 1
+    min_int = -(2 ** (n_bit - 1))
+    # TODO: currently we don't know how to express torch.int4, we'll
+    # add torch.int4 to core later
+    from torchao._executorch_ops import _quantized_decomposed_quantize_per_channel_group_wrapper
+    w_int8 = _quantized_decomposed_quantize_per_channel_group_wrapper(
+        w, scales, zeros, min_int, max_int, torch.int8, group_size
+    )
 
-        return w_int8, scales, zeros
+    return w_int8, scales, zeros
 
 
-    def down_size(size):
-        assert size[-1] % 2 == 0, f"{size} last dim not divisible by two"
-        return (*size[:-1], size[-1] // 2)
+def down_size(size):
+    assert size[-1] % 2 == 0, f"{size} last dim not divisible by two"
+    return (*size[:-1], size[-1] // 2)
 
 
-    def up_size(size):
-        return (*size[:-1], size[-1] * 2)
+def up_size(size):
+    return (*size[:-1], size[-1] * 2)
 
 
-    quantized_decomposed_lib.define("pack_int4_from_int8(Tensor int8_data) -> Tensor")
+quantized_decomposed_lib.define("pack_int4_from_int8(Tensor int8_data) -> Tensor")
 
 
-    @impl(quantized_decomposed_lib, "pack_int4_from_int8", "CompositeExplicitAutograd")
-    def pack_int4_from_int8(int8_data: torch.Tensor) -> torch.Tensor:
-        # converting to uint8 for operations
-        shape = int8_data.shape
-        assert shape[-1] % 2 == 0
-        int8_data = int8_data.contiguous().view(-1)
-        return (int8_data[::2] << 4 | int8_data[1::2]).view(down_size(shape))
+@impl(quantized_decomposed_lib, "pack_int4_from_int8", "CompositeExplicitAutograd")
+def pack_int4_from_int8(int8_data: torch.Tensor) -> torch.Tensor:
+    # converting to uint8 for operations
+    shape = int8_data.shape
+    assert shape[-1] % 2 == 0
+    int8_data = int8_data.contiguous().view(-1)
+    return (int8_data[::2] << 4 | int8_data[1::2]).view(down_size(shape))
 
 
-    quantized_decomposed_lib.define("unpack_int4_to_int8(Tensor int8_data) -> Tensor")
+quantized_decomposed_lib.define("unpack_int4_to_int8(Tensor int8_data) -> Tensor")
 
 
-    @impl(quantized_decomposed_lib, "unpack_int4_to_int8", "CompositeExplicitAutograd")
-    def unpack_int4_to_int8(int8_data: torch.Tensor) -> torch.Tensor:
-        """ Get the original weight from the normalized float weight format"""
-        # since we are using int8 we will decode 2 entries per byte
-        # Shift elements down 4 and select out the bottom 4 bits
-        shape = int8_data.shape
-        first_elements = (int8_data >> 4).to(torch.int8)
-        second_elements = (int8_data & 0b1111).to(torch.int8)
-        return torch.stack([first_elements, second_elements], dim=-1).view(up_size(shape))
+@impl(quantized_decomposed_lib, "unpack_int4_to_int8", "CompositeExplicitAutograd")
+def unpack_int4_to_int8(int8_data: torch.Tensor) -> torch.Tensor:
```

Comment:
btw we could probably remove these, I think these are not used by executorch right now

Reply:
Ok, I'll do that in a follow up.

Reply:
please go ahead with the follow up, I'm planning to test out ghstack again so we can make this process easier in the future